addons/io_hubs_addon/third_party/recast.py

```diff
@@ -0,0 +1,636 @@
+        libpath = os.path.abspath(addon_prefs.recast_lib_path)
+        libpathr = libpath.replace("\\", "/")
+        if not os.path.exists(libpathr):
+            self.report({'ERROR'}, 'File not exists: %s\n' % libpathr)
+            return {'FINISHED'}
+
+        verts, tris = extractTriangulatedInputMesh(context)
+        vertsCount = len(verts)
+        trisCount = len(tris)
+        nverts = (int)(len(verts) / 3)
+        ntris = (int)(len(tris) / 3)
+        recastData = recastDataFromBlender(context.scene)
+        if context.scene.recast_navmesh.auto_cell:
+            recastData.cellsize = get_auto_cell_size(context)
+
+        prevWorkingDir = os.getcwd()
+        nextWorkingDir = os.path.dirname(libpathr)
+        os.chdir(nextWorkingDir)
+        try:
+            recast = ctypes.CDLL(libpathr)
+        except OSError as e:
+            tracebackStr = traceback.format_exc()
+            self.report(
+                {'ERROR'},
+                'Failed to load shared library: %s\nPath to shared library: %s\n\nTraceback: %s' %
+                (str(e),
+                 libpathr, tracebackStr))
+            os.chdir(prevWorkingDir)
+            return {'FINISHED'}
+
+        os.chdir(prevWorkingDir)
+
+        pmesh = recast_polyMesh_holder()
+        dmesh = recast_polyMeshDetail_holder()
+        nreportMsg = 128
+        reportMsg = ctypes.create_string_buffer(b'\000' * nreportMsg)     # 128 chars mutable text
+        recast.buildNavMesh.argtypes = [
+            ctypes.POINTER(RecastData),
+            c_int, c_float * vertsCount, c_int, c_int * trisCount, ctypes.POINTER(recast_polyMesh_holder),
+            ctypes.POINTER(recast_polyMeshDetail_holder),
+            ctypes.c_char_p, c_int]
+        recast.buildNavMesh.restype = c_int
+        recast.freeNavMesh.argtypes = [
+            ctypes.POINTER(recast_polyMesh_holder),
+            ctypes.POINTER(recast_polyMeshDetail_holder),
+            ctypes.c_char_p, c_int]
+        recast.freeNavMesh.restype = c_int
+
+        ok = recast.buildNavMesh(recastData, nverts, (c_float * vertsCount)(* verts), ntris,
+                                 (c_int * trisCount)(*tris), pmesh, dmesh, reportMsg, nreportMsg)
+        print("Report msg: %s" % reportMsg.raw)
+        if not ok:
+            self.report({'ERROR'}, 'buildNavMesh C++ error: %s' % reportMsg.value)
+
+        if not dmesh.dmesh:
+            self.report({'ERROR'}, 'buildNavMesh C++ error: %s' % 'No recast_polyMeshDetail')
+        else:
+            # print("ABC %i" % pmesh.pmesh.contents.nverts)
+            # dmeshv1 = dmesh.dmesh.contents.verts[0]
+            # print("dmeshv1 %f" % dmeshv1)
+
+            createMesh(context, dmesh, obj=navMesh)
+
+        # what was allocated in C/C++ should be also deallocated there
+        recast.freeNavMesh(pmesh, dmesh, reportMsg, nreportMsg)
+
+        bpy.ops.object.select_all(action='DESELECT')
+        for obj in selected_objects:
+            obj.select_set(True)
```

Comment:
I think the active object should also be saved/restored as well as the selected objects?